src/pybind/mgr/mgr_module.py

```diff
@@ -200,6 +204,66 @@ def get_default_choose_args(dump):
         return dump.get('choose_args').get(CRUSHMap.DEFAULT_CHOOSE_ARGS, [])
 
 
+class CLICommand(object):
+    COMMAND_LIST = []
+
+    def __init__(self, prefix, args=None, desc=None, perm=None):
+        self.prefix = prefix
+        self.args = args if args else ""
+        self.args_dict = {}
+        self.desc = desc if desc else ""
+        self.perm = perm if perm else "rw"
+        self.func = None
+        self._parse_args()
+
+    def _parse_args(self):
+        if not self.args:
+            return
+        args = self.args.split(" ")
+        for arg in args:
+            arg_desc = arg.strip().split(",")
+            arg_d = {}
+            for kv in arg_desc:
+                k, v = kv.split("=")
+                if k != "name":
+                    arg_d[k] = v
+                else:
+                    self.args_dict[v] = arg_d
+
+    def __call__(self, func):
+        self.func = func
+        setattr(self.func, '_cli_cmd_', True)
+        self.COMMAND_LIST.append(self)
+        return self.func
+
+    def call(self, mgr, cmd_dict):
+        kwargs = {}
+        for a, d in self.args_dict.items():
+            if 'req' in d and d['req'] == "false" and a not in cmd_dict:
+                continue
+            kwargs[a.replace("-", "_")] = cmd_dict[a]
+        return self.func(mgr, **kwargs)
+
+    @classmethod
+    def register_commands(cls, mgr_module):
+        for cmd in cls.COMMAND_LIST:
+            mgr_module.COMMANDS.append({
+                'cmd': '{} {}'.format(cmd.prefix, cmd.args),
+                'desc': cmd.desc,
+                'perm': cmd.perm
+            })
+
+
+class CLIReadCommand(CLICommand):
```

Comment:
Do we need a class here?
```python
def CLIReadCommand(prefix, args=None, desc=None):
    CLICommand(prefix, args, desc, "r")
```

Reply:
Nope. This is the consequence of my more object-oriented experience. I'll follow your suggestion.